Ladder step2

clown0425 opened the pull request.

clown0425 committed:
유저의 입력을 저장하는 프로퍼티 생성

clown0425 committed:
입력받은 값의 옵셔널을 벗겨내고
string을 입력했는지 확인하고
맞다면 string으로 변환된 값을 리턴핟다
string으로 변환할수 없다면
다시 입력 받을 수 있도록한다

clown0425 committed:
string을 Int로 변환한다
만약 변환에 실패 한다면
처음부터 다시 입력받게 한다

clown0425 committed:
입력받은 값을 저장할 프로퍼티 생성

clown0425 committed:
플레이어수를 기준으로
생성할 사다리배열의 가로 길이 계산후 저장

clown0425 committed:
모든 아이템이 공백인 이차원배열 생성
laz를 붙여야만 오류가 나타나지 않는데 왜 그러는지 정확하게 모르겠다
(검색을 해보니 호출 중 초기화 되지 않은 프로퍼티를 사용할 수도 있기 떄문에 초기화를 먼저 완료하지 않으면 컴파일 에러가 발생한다고 한다)
????

clown0425 committed:
가로 사다리개수를 랜덤으로 생성해 주는 함수
연속되지 않게 해야 하기 때문에
겹치지 않게 놓을수 있는 최악의 경우 이하의 숫자만 생성하게 했다
이부분을 개선할수 있는지 생각해봐야 한다

clown0425 committed:
2차원 배열에서 랜덤으로 가로와 세로 좌표를 생성후 저장

clown0425 committed:
세로 사다리 입력 메서드

clown0425 committed:
가로 사다리를 조건에 맞게 입력 하는 메서드
메서드를 더 작게 나누고 싶은데 방법이 떠오르지 않는다

clown0425 committed:
결과를 프린트

clown0425 committed:
초기화 형식 작성

clown0425 committed:
플레이어수와 사다리높이를 입력받아 리턴

clown0425 committed:
사다리 입력받은 수를 기준으로 사다리그리기

clown0425 committed:
입력과 그리기를 묶어주기

clown0425 committed:
실행

godrm commented on LadderGame/LadderGame/playLadderGame.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+struct play {
```
타입을 만들때는 대문자로 시작하세요.

godrm commented:
근데 4-2를 하는데 일부러 struct를 사용한건가요?
4-3에서 struct 를 하는게 포함되어 있는데 그걸 고려한건가요?

godrm commented on LadderGame/LadderGame/playLadderGame.swift:
```diff
@@ -0,0 +1,32 @@
+        print("플레이어 수를 입력하세요")
+        let playerNumber = input.convertInt(input.removeOptional(input.requestNumber))
+        print("사다리 높이를 입력하세요")
+        let ladderHeight = input.convertInt(input.removeOptional(input.requestNumber))
```
여기 print() 안내 문구를 포함, input.convertInt() 함수 호출까지도 일반화해서 공통 함수로 분리해보세요.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -0,0 +1,33 @@
+        if let convertedInt = Int(convertedString){
+            return convertedInt
+        }
+        print("입력할 값을 다시 확인하세요")
```
무언가 판단하거나 검증하는 로직을 처리한 코드에 print() 처럼 직접 안내문구를 출력하는 걸 포함하면 안됩니다.
대신 정확한 의도된 내용을 return 해서 이 함수를 호출한 상위 로직에서 흐름을 제어해서 출력을 할지, 다시 호출할지 변경할 수 있어야 합니다.

godrm commented on LadderGame/LadderGame/playLadderGame.swift:
```diff
@@ -0,0 +1,32 @@
+        var test = LadderGameValue(playerNumber:input[0],ladderHeight:input[1])
+        test.addHeight()
+        test.addWidth()
+        test.resultprint()
```
- 매개변수 input은 어떤 역할이고, 어떤 값을 담고 있나요? 전혀 정보가 없습니다. input 이라는 변수명을 구체적인 정보를 포함하는 단어로 바꿔보세요.
- test 라는 변수명은 의도를 표현하기 부족한 이름이네요. 역할이 뭔지 명확하게 써주세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -8,3 +8,89 @@
+    mutating func resultprint(){
+        for i in ladders{
+            print(String(i))
+        }
```
이 LadderGameValue 구조체는 어떤 역할을 기대한건가요? 
LadderGame을 위한 값을 포함하는 구조체인가? 
혹은 사다리게임을 구현하기 위한 사다리 생성을 포함하는 로직인가요?
심지어 출력을 하는 resultPrint() 함수도 포함하고 있나요? 
역할을 구분할 때는 한 가지 책임과 역할만 하도록 분리해야 합니다.

godrm reviewed (commented):
구조체를 하기 전에 함수 분리를 먼저 연습했어야 했는데 
구조체를 포함하고 있어서 구조체 내용까지 리뷰를 했습니다. 
4-2 내용과 4-3 내용이 혼합되어 있는 느낌입니다. 다음 피드백을 확인해보세요.

clown0425 committed:
입력을 요구하는 함수를 공통 함수로 분리

clown0425 committed:
매개변수 input을 ladderInitializationMamber로 변경 초기화에 필요한 값을 전달하는게 목적
userInputRequest() 함수 역시 메세지가 명확하지 않아 ladderGameBasicValueRequest로 변경
test 변수명 코딩중 실험해볼 목적으로 사용하고 수정하지 않음 ladder로 변경

clown0425 committed:
사다리 생성에 필요한 값을 저장하는 구조체
사다리를 생서하는 동작을 하는 구조체
사다리를 출력하는 구조체
역할을 세분화해 구현

clown0425 committed:
판단하거나 검증하는 로직에서 모든걸 처리하지 않고
상위 로직으로 넘겨서 처리하는 것으로 바꾸었다

우선 사용자로 부터 받은 정보가 원하는 정보인지 확인하고
맞다면 받은값 그대로를 리턴 아니라면 공백이나 0을 리턴 함으로써
원하는 정보의 여부를 판단하여
상위로직에서 흐름을 제어하도록 해보았다
이렇게 처리하는게 맞는지 아직 확실히 모르겠다

clown0425 committed:
고쳤음에도 불구 여전히 판단하거나 검증하는 로직에서 모든걸 처리하고 있음
검증하는 로직에서는 정확히 실패나 성공한 값을 리턴하고
상위로직에서 리턴값을 토대로 다른 함수를 호출하거나 성공값을 리턴하는 형태로 수정

clown0425 committed:
가로 사다리를 추가하는 함수 수정
하나의 함수에서 너무 많은 역할을 하고 있어서
하나의 역할만 할수있게 나누어 보았습니다

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+import Foundation
+
+struct AddLadder {
+    
```
구조체 타입 이름은 명사형으로 작성하는 게 좋습니다. 
`AddLadder`는 어떤 의도에 대한 이름인가 궁금합니다. 
구조체라면 함수들 동작을 추상화해서 부를 수 있는 이름이어야 할 것 같습니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+    }
+    
+    func widthCoordinate(_ playerNumber:Int)->Int{
+        return (Int(arc4random_uniform(UInt32(playerNumber-1))) * 2) + 1
```
- UInt32 타입을 사용하는 arc4random_uniform() 보다는 Int.random() 함수를 활용하세요.
- width를 계산하는데 `((playerNumber-1)*2)+1 ` 이렇게 복잡한 계산이 되어 버린건 아마도 구조 때문일 겁니다. 
사다리를 구성하는 기본 데이터 구조를 딱 사다리 발판만 저장하는 구조로 만들어보세요.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+        var ladders = ladders
+        for i in 0..<playerNumber{
+            for j in 0..<ladderHeight{
+                ladders[j][i*2] = "ㅣ"
```
특히 "ㅣ" 는 사다리 출력을 위한 형식일 뿐입니다.
형식은 언제든지 바뀔 수 있습니다. 이런 출력에 대한 값을 미리 데이터 구조에 저장할 필요는 없습니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+    
+    func Randomcordinates(ladderHeight:Int,playerNumber:Int) -> Array<Int> {
+        let height = heightCoordinate(ladderHeight) ,width = widthCoordinate(playerNumber)
+        return [height,width]
```
함수 이름이 `Randomcordinates` 대문자가 포함되어 있네요.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+        return [height,width]
+    }
+    
+    func isEmpty(ladders:[[Character]],coordinates:Array<Int>,empty:Character,full:Character) -> Character {
```
coordinates가 좌표를 의미하나요? x,y 두 값만 써도 되는데 배열을 사용하는 건 오버인 것 같습니다.
배열을 사용할 경우는 오히려 배열에 [0] 값이 없는 경우, [1] 값이 없는 경우 ... 더 많이 들어있는 경우까지 예외 상황이 많아집니다. 배열을 사용한다는 것은 그만큼 값이 더 늘어갈 수 있다는 구조에서 선택하는 게 좋습니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+        guard right == empty && left == empty else{
+            return full
+        }
+        return empty
```
구조체를 사용하는 이유 중에 하나는 매개 변수로 자주 사용하는 값들은 내부 속성으로 사용하는 장점이 있기 때문입니다. 
여기 함수들은 계속해서 반복하는 값을 사용하고 있네요. 그렇다는 얘기는 하나의 구조체로 표현할 수도 있다는 의미기도 합니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,124 @@
+        ladders = ladderDraw.addHeight(playerNumber: ladder.playerNumber, ladderHeight: ladder.ladderHeight, ladders: ladders)
+        ladders = ladderDraw.addWidth(widthLadderNumber: ladder.widthLadderNumber, ladders: ladders, ladderHeight: ladder.ladderHeight, playerNumber: ladder.playerNumber, ladderWidthLength: ladder.ladderWidthLength)
+        return ladders
+    }
```
Function 이라는 타입 이름은 무엇을 의도한 건가요? 
내부 기능과 역할을 추상화한 단어가 Function 이면 그냥 기능인가요? 함수인가요?

godrm commented on LadderGame/LadderGame/test.swift:
```diff
@@ -0,0 +1,92 @@
+            index += 1
+        }
+        return ladders
+    }
```
여기 test.swift 파일에 있는 Test 구조체는 어떤 역할과 책임인지 모르겠습니다.
위에 구현한 다른 객체와 중복인가요? 그냥 임시로 만든거라면 커밋에 포함하면 안됩니다.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+    }
+}
+
+struct InputMamber {
```
`InputMamber` 사소한 디테일이지만 오타가 발생하면 그 코드를 읽는 사람에게 전혀 다른 의미가 될 수 있습니다.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+        if convertedString == fail {
+            return fail
+        }
+        return convertedString
```
isString() 에서 `is-` 는 스위프트 API 가이드라인 기준에 따르면 true/false 로 구분이되는 Bool 타입에만 사용합니다. 어떤 변환하는 문자열로 리턴하는 거라면 그 변환하는 문자열을 표현하는게 좋겠습니다.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+            print("입력한 값을 다시 확인하세요, 정수를 입력해야 합니다")
+            return userInputNumber()
+        }
+        return convertedInt
```
여기까지 변환하는 함수들 모두 마찬가지입니다. 
`comvertInt` 이 함수에도 오타가 포함되어 있네요.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+    func userInputNumber()->Int{
+        let convertedString = convertString()
+        let convertedInt = comvertInt(convertedString)
+        return convertedInt
```
userInputNumber() 함수는 comvertInt()를 호출하고
comvertInt()에서는 userInputNumber()를 호출하고 있습니다.
이렇게 상호 호출하는 것도 재귀 호출이 됩니다. 이런 구조는 흐름제어를 할 수 없기 때문에 지양하세요.
검증하는 함수가 있다면 그 결과만 리턴하고, 그 함수를 호출한 상위 함수에서 다시 호출하는 구조로 만드세요.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+    
+    func ladderHeight()->Int{
+        print("사다리 높이를 입력하세요")
+        let ladderHeight = ladderHeightCheck(isoverOne(userInputNumber()))
```
억지로 한 줄로 줄여서 작성하지 마세요. 
당분간은 풀어서 작성하세요.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -15,19 +15,141 @@ struct UserInputManage{
+        return ladderHeight
+    }
+    
+    func LadderGameBasicValueInitMamber()->Array<Int>{
```
여기도 함수 이름이 대문자로 시작하고 있습니다.

godrm reviewed (commented):
스탭 4-2는 4-1에서 작성한 함수를 분리하는 것인데
4-3 요구사항에 있는 구조체까지 하면서 뒤죽박죽인 상태네요.
구조체를 작성하기 위해서 함수 분리가 안되어 있으면 구조체 역할 분리가 지금처럼 얽히게 됩니다. 

최상위 모듈부터 차례차례 역할과 책임이 더 구체적인 단계로 나눠지고, 
상위 함수는 하위 객체/함수를 조합해서 작업을 처리해야 합니다. 
피드백을 확인해보세요.

clown0425 committed:
플레이어 숫자와 사다리 높이를 입력받는 부분을 수정
재귀적 호출이 아니라 상위함수에서 반복문을 통해 동작하는 방법으로 바꿈
기타 변수명 수정

clown0425 committed:
데이터구조를
가로사다리는 불타입으로 저장하고 그것을 기준으로 출력한다
필요한 값을 분리하지 않고 같은 구조체로 합침

clown0425 committed:
함수의 역할을 더 작게 분리하였음
판단을 요하는 함수는 불린 값을 리턴하게 하였고
상위 함수에서 불린값을 받아 다음 처리를 하게 하였다
(isString,isInt,isOverOne)

clown0425 committed:
사다리게임 실핼하는 함수(playLadderGame)수정

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -29,127 +28,83 @@ struct UserInputManage{
+            print("최소 플레이어수는 2입니다, 플레이어수를 다시 입력해주세요")
+            userInputInt()
+            isoverOne = self.isOverOne()
+        }
```
함수 내부에서 반복문으로 흐름을 반복하도록 하지마세요. 대신 리턴해서 최상위 함수에서 다시 호출하는게 좋습니다.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -0,0 +1,136 @@
+    mutating func allWidthLadderDraw() {
+        for i in 0..<ladderHeight {
+            for j in 0..<playerNumber-1{
+                drawWidthLadder(i,j)
```
이중 for 문에서 하위 루프를 별도 함수로 분리해보세요. 
들여쓰기가 깊어지지 않도록 함수를 분리하는 연습중입니다.

godrm reviewed (commented):
함수 분리 연습만 하고 마무리하면 될 것 같습니다.

clown0425 committed:
함수 안에서 흐름을 반복하지 않게 분리해 보았습니다

clown0425 committed:
for문을 분리하였습니다
함수내부에서 반복문으로 흐름을 반복하지 않도록 분리하였습니다

clown0425 committed:
userInputInt함수 흐름 반복하지 않게 분리

clown0425 committed:
drawWidthLadder 함수 이중for문 분리

clown0425 commented:
1.userInputInt, Number, markWidthAsmuchWidthLadderNumber, findNotContinueCoordinate
위의 함수 모두 흐름을 반복하지 않게 분리 해 보았습니다
제가 코멘트를 제대로 이해하고 분리한 건지 잘 모르겠습니다

2.checkNumber 
거의 똑같은 기능을 하는 플레이어를 입력받는 함수와 사다리 높이를 입력받는 함수를
전달받는 인자에 따라 역할을 할수있도록 공통으로 분리해 구현했습니다

3.allHeightLadderDraw, allWidthLadderDraw
이중 for문을 분리해서 구현했습니다

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -74,32 +80,32 @@ struct UserInputManage{
+            number = self.checkNumber(RequestMassage)
         }
+            while number == false
     }
```
- 함수 이름은 소문자로 시작하고, 동사형으로 행위를 표현하세요.

godrm commented on LadderGame/LadderGame/ladderDraw.swift:
```diff
@@ -80,12 +80,20 @@ struct Draw {
+            NotContinueCoordinate = checkNotContinueCoordinate()
         }
-            while isContinue() == false
+            while NotContinueCoordinate == false
```
변수 이름도 소문자로 작성하세요. 
스위프트 API 디자인 가이드라인 문서를 꼭 읽어보세요. Bool 변수인 경우는 어떻게 쓰는지 예시도 있습니다.

godrm commented on LadderGame/LadderGame/userInput.swift:
```diff
@@ -57,13 +57,19 @@ struct UserInputManage{
-            userInputString()
-            isInt = self.isInt()
+            checkUserInputInt = self.checkUserInputInt()
         }
```
반복문을 특정한 하위 함수 내에서 사용하지 않고, 상위 모듈에서 결과를 받아서 처리하도록 개선하세요.

clown0425 replied:
생각을 해봐도 어떻게 해야 되는지 잘 모르겠습니다 
어쨌든 조건에 맞을 때 까지 반복 해야 하는데 재귀도 반복문도 안쓰고 어떻게 가능할지 방법을 모르겠어요 ㅜㅜ
조금만더 힌트를 주세요~~

godrm replied:
반복문을 사용하는 곳을 조정해보세요. 
하위 모듈에서는 사용하지 말고, 상위 모듈에서 처리하자는 겁니다.
이 프로그램에서 최상위 모듈은 main.swift 에 작성하는 최상위 함수를 말합니다.

clown0425 committed:
변수를 생성하고 저장하는 과정이 필요 없는 것 같아 삭제

clown0425 committed:
함수이름을 기존 명사형에서 동사형 으로 변경

clown0425 committed:
convertedString 위치 바꿈

clown0425 committed:
convertedInt 위치 바꿈

clown0425 committed:
removeOptional 인자를 전달 받지 않게 변경

clown0425 committed:
하나의 함수가 하나의 역할만을 하게 수정
반복은 상위 함수에서 하는 것으로 수정

clown0425 committed:
변수와 함수의 이름을 수정하고
상위함수에서 반복하게함

clown0425 committed:
userInput 구조체를 수정했기 때문에 그에 따라 약간의 수정이 있었음

clown0425 committed:
ladderBoardInitNumber함수 개선 if문을 나열하지 않고 반복할수 있게 변경
하지만 중간에 실패함을 알아도 끝까지 진행해야 한다는 단점이 있음

clown0425 committed:
ladderBoardInitNumber 함수 분리

clown0425 committed:
markAsmuchWidthLadderNumber함수 분리

clown0425 committed:
findCorrectCoordinate함수 분리

clown0425 commented:
전제적으로 이름을 수정하였습니다
함수하나는 하나의 역할만 할수있게 분리해 보았습니다
반복문을 상위모듈에서 처리하게 수정해 보았습니다
(제가 이해한게 맞는지 아직 확신이 서지 않습니다)

godrm commented:
함수 분리에 대해 고민해봤으니 이제 구조체로 공통적인 데이터와 동작을 합쳐보세요.
